netty/src/main/java/io/grpc/netty/NettyClientHandler.java

```diff
@@ -350,6 +351,15 @@ private void createStream(CreateStreamCommand command, final ChannelPromise prom
     final NettyClientStream stream = command.stream();
     final Http2Headers headers = command.headers();
     stream.id(streamId);
+
+    if (connection().goAwayReceived() && streamId > connection().local().lastStreamKnownByPeer()) {
```

Comment:
I'm not confident `goAwayReceived` is the only case that needs to be handled. It seems like any case `goAwayStatus != null`, then we should probably close the stream. For example, if `onConnectionError` was called and then `channelInactive`, then all streams would be closed with a helpful status. A new stream though would not trigger the condition here and I would expect to fail with something unhelpful, like `ClosedChannelException`.

However, when trying to verify that `ClientTransport.Listener.transportShutdown` is called anytime `goAwayReceived` is set (since we want to drive away new calls; not just continually kill them), I can't find any such calls to `transportShutdown` (see `NettyClientTransport`) except for when the connection actually closes.

If that's actually the case, then we aren't handling GOAWAY correctly: we cancel all calls created after GOAWAY was received and before all existing streams complete (which has no time-bound; it could be minutes or hours). The reason we may not have noticed is that, GFE has had weak GOAWAY behavior itself in that it closed the connection immediately.